code/modules/paperwork/contract.dm

```diff
@@ -0,0 +1,293 @@
+				else
+					if(signed)
+						to_chat(user,"<span class='notice'>This contract has already been signed.  It may not be signed again.</span>")
+					else
+						to_chat(user,"<span class='notice'>You quickly scrawl your name on the contract</span>")
+						if(FulfillContract(target.current, blood)<=0)
+							to_chat(user,"<span class='notice'>But it seemed to have no effect, perhaps even Hell itself cannot grant this boon?</span>")
+						return 1
+			else
+				to_chat(user, "<span class='notice'>This devil already owns your soul, you may not sell it to them again.</span>")
+		else
+			to_chat(user,"<span class='notice'>Your signature simply slides off the sheet, it seems this contract is not meant for you to sign.</span>")
+	else
+		to_chat(user, "<span class='notice'>You don't know how to read or write.</span>")
+	return 0
+
+
+
+/obj/item/weapon/paper/contract/infernal/revive/attack(mob/M, mob/living/user)
+	if (target == M.mind && M.stat == DEAD && M.mind.soulOwner == M.mind)
+		var/mob/living/carbon/human/H = M
+		var/mob/dead/observer/ghost = H.get_ghost()
+		var/response = "No"
+		if(ghost)
+			ghost.notify_cloning("A devil has offered you revival, at the cost of your soul.",'sound/effects/genetics.ogg', H)
+			response = alert(ghost, "A devil is offering you another chance at life, at the price of your soul, do you accept?", "Infernal Resurrection", "Yes", "No")
+			if(!ghost)
+				return		//handle logouts that happen whilst the alert is waiting for a response.
+		else
+			response = alert(target.current, "A devil is offering you another chance at life, at the price of your soul, do you accept?", "Infernal Resurrection", "Yes", "No")
+		if(response == "Yes")
+			H.revive(1,0)
+			add_logs(user, H, "infernally revived via contract")
+			user.visible_message("<span class='notice'>With a sudden blaze, [H] stands back up.</span>")
+			H.fakefire()
+			FulfillContract(H, 1)//Revival contracts are always signed in blood
+			spawn(5)
+				H.fakefireextinguish(TRUE)
+	else
+		..()
+
+
+/obj/item/weapon/paper/contract/infernal/proc/FulfillContract(mob/living/carbon/human/user = target.current, blood = 0)
+	signed = 1
+	if(user.mind.soulOwner != user.mind && user.mind.soulOwner.devilinfo) //They already sold their soul to someone else?
+		user.mind.soulOwner.devilinfo.remove_soul(user.mind) //Then they lose their claim.
+	user.mind.soulOwner = owner
+	user.hellbound = contractType
+	user.mind.damnation_type = contractType
+	owner.devilinfo.add_soul(user.mind)
+	update_text(user.real_name, blood)
+	to_chat(user, "<span class='notice'>A profound emptiness washes over you as you lose ownership of your soul.</span>")
+	to_chat(user, "<span class='boldnotice'>This does NOT make you an antagonist if you were not already.</span>")
+	return 1
+
+/obj/item/weapon/paper/contract/infernal/power/FulfillContract(mob/living/carbon/human/user = target.current, blood = 0)
+	if(!user.dna)
+		return -1
+	user.mutations.Add(HULK)
```

Comment:
I had to add an update icons after it, good catch.  The contract now properly turns you green.